### PR TITLE
PS-5221 : Page Cleaner will sleep for long time if clock changes

### PR DESCRIPTION
**Pull request by george-lorch**
- Added monotonic clock source for transient comparisons.

- Changed InnoDB page flusher, lru, and bp load throttle to use new monotonic
  clock source.

### CODE REVIEW

**Comment by george-lorch** [comment]
https://ps.cd.percona.com/view/5.7/job/percona-server-5.7-param/97/

**Comment by laurynas-biveinis** [comment]
One thing that I missed in the original discussion is that we should fix 5.6 too, right?

File: storage/innobase/buf/buf0dump.cc [hunk 1]
```diff
@@ -451,8 +451,9 @@ buf_load_throttle_if_needed(
 
 	/* There has been other activity, throttle. */
 
-	ulint	now = ut_time_ms();
-	ulint	elapsed_time = now - *last_check_time;
+	const ut_monotonic_time	now = ut_monotonic_time_ms();
+	ut_monotonic_time	elapsed_time;
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0dump.cc]: elapsed_time staying as ulint would result in a smaller diff?

File: storage/innobase/buf/buf0flu.cc [hunk 2]
```diff
@@ -3290,14 +3291,14 @@ DECLARE_THREAD(buf_flush_page_cleaner_coordinator)(
 			/* Request flushing for threads */
 			pc_request(ULINT_MAX, lsn_limit);
 
-			ulint tm = ut_time_ms();
+			ut_monotonic_time tm = ut_monotonic_time_ms();
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0flu.cc]:
```suggestion
			const ut_monotonic_time tm = ut_monotonic_time_ms();
```

File: storage/innobase/buf/buf0flu.cc [hunk 3]
```diff
@@ -3333,14 +3334,14 @@ DECLARE_THREAD(buf_flush_page_cleaner_coordinator)(
 			/* Request flushing for threads */
 			pc_request(n_to_flush, lsn_limit);
 
-			ulint tm = ut_time_ms();
+			ut_monotonic_time tm = ut_monotonic_time_ms();
```
Comment (laurynas-biveinis) [inline storage/innobase/buf/buf0flu.cc]:
```suggestion
			const ut_monotonic_time tm = ut_monotonic_time_ms();
```

File: storage/innobase/include/ut0ut.h [hunk 4]
```diff
@@ -261,6 +261,22 @@ ulint
...
+code and reveal any changes that might later be introduced by upstream.
+*/
+struct ut_monotonic_time
+{
+	ulint ms;
+};
+
+ut_monotonic_time
+ut_monotonic_time_ms(void);
+/*============*/
```
Comment (laurynas-biveinis) [inline storage/innobase/include/ut0ut.h]:
```suggestion
```

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by george-lorch** [comment]
Hmm 5.6?  That is entirely your call to make @laurynas-biveinis  I can't imagine it being difficult to back port this onto a GCA point.

File: storage/innobase/buf/buf0dump.cc (hunk 1 above)
Comment (george-lorch) [reply storage/innobase/buf/buf0dump.cc]: True, it doesn't look like it currently would be a threat from an upstream change that needs triggering on a merge.

**Comment by george-lorch** [comment]
Re-pushed due to 5.6 and GCA merge forward.  New jenkins with jobs that appear to have been aborted due to spot instance demand https://ps.cd.percona.com/view/5.7/job/percona-server-5.7-param/111/.  Going to go ahead and merge as requested changes have been made.

**Pull request merged**